core/domain/stats_services.py

```diff
@@ -273,8 +278,38 @@ def advance_version_of_exp_stats(
         exp_stats.state_stats_mapping[state_name] = (
             stats_domain.StateStats.create_default())
 
+    # If state names were swapped, both the new and old names will be present in
```

Comment:
The usual way to handle this kind of issue is to create a completely new object and populate it from scratch, rather than mutate the original object in-place. Otherwise, you fall prey to other issues like a three-way swap (A --> B, B --> C, C --> A etc.) and there is no guarantee this is comprehensive.

Reply:
Oh I see what you mean. I'll try to fix this.

Reply:
Hang on -- the logic in the approach I described should not depend on whether states were swapped or not. It should simply be something like this:

```
new_state_stats_mapping = {}
for each state:
    new_state_stats_mapping[new_name_of_state] = ...
exp_stats.state_stats_mapping = new_state_stats_mapping
```

Reply:
Ah, but before this we have two loops to handle state additions and state deletions. Creating a `new_state_stats_mapping` dict and replacing `exp_stats.state_stats_mapping` after `exp_versions_diff.new_to_old_state_names` loop might result in some loss in data.

Reply:
Ah yes. Sorry I forgot a step, we would also need to maintain a list of added states and process them after the for loop. And if we delete a state, the "new state name" for that state is None.

Reply:
Done.